File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    const visited = Array.from({length: dungeons.length}, ()=>0);
    
    function backtracking(curK, cnt){
        let max = cnt;
        for(let i=0; i<visited.length; i+=1){
            if(visited[i]===1 || dungeons[i][0]>curK) continue;
            
            visited[i] = 1;
            max = Math.max(max, backtracking(curK-dungeons[i][1], cnt+1));
            visited[i] = 0;
        }
        return max;
    }
    
    const max = backtracking(k, 0);
    return max;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    int answer;

    public int solution(int k, int[][] dungeons) {
        answer = 0;
        dfs(k, new boolean[dungeons.length], dungeons, 0);
        return answer;
    }

    public void dfs(int hp, boolean[] visited, int[][] dungeons, int cnt) {
        answer = Math.max(answer, cnt);
        for (int i = 0; i < dungeons.length; i++) {
            if (visited[i]) continue;
            if (hp < dungeons[i][0]) continue;

            visited[i] = true;
            dfs(hp - dungeons[i][1], visited, dungeons, cnt + 1);
            visited[i] = false;
        }
    }
}